examples/epocher/src/application/actor.rs

```diff
@@ -0,0 +1,260 @@
+use super::ingress::{Mailbox, Message};
+use crate::{
+    orchestrator::EpochCert,
+    types::{
+        block::{Block, GENESIS_BLOCK, GENESIS_ROUND},
+        epoch,
+    },
+};
+use commonware_consensus::{
+    marshal,
+    threshold_simplex::types::{Context, Finalization},
+    types::Round,
+    Block as _, Reporter,
+};
+use commonware_cryptography::{
+    bls12381::primitives::variant::MinSig, sha256::Digest as Sha256Digest, Committable,
+};
+use commonware_macros::select;
+use commonware_runtime::{Handle, Metrics, Spawner, Storage};
+use commonware_utils::futures::ClosedExt;
+use futures::{
+    channel::{mpsc, oneshot},
+    StreamExt,
+};
+use rand::Rng;
+use std::collections::BTreeMap;
+use tracing::{debug, info, warn};
+
+/// Application actor for a single-network epocher.
+///
+/// This actor is responsible for proposing and verifying blocks.
+/// It also tracks the finalized blocks and reports to the orchestrator when the epoch is complete.
+pub struct Application<R: Rng + Spawner + Metrics + Storage> {
+    context: R,
+    mailbox: mpsc::Receiver<Message<Sha256Digest>>,
+    marshal: marshal::Mailbox<MinSig, Block>,
+
+    /// Responders for block heights.
+    responders: BTreeMap<u64, Vec<oneshot::Sender<Sha256Digest>>>,
+}
+
+impl<R: Rng + Spawner + Metrics + Storage> Application<R> {
+    pub fn new(
+        context: R,
+        mailbox_size: usize,
+        marshal: marshal::Mailbox<MinSig, Block>,
+    ) -> (Self, Mailbox<Sha256Digest>) {
+        let (sender, mailbox) = mpsc::channel(mailbox_size);
+        (
+            Self {
+                context,
+                mailbox,
+                marshal,
+                responders: BTreeMap::new(),
+            },
+            Mailbox::new(sender),
+        )
+    }
+
+    pub fn start<O: Reporter<Activity = EpochCert>>(mut self, orchestrator: O) -> Handle<()> {
+        self.context.spawn_ref()(self.run(orchestrator))
+    }
+
+    async fn run<O: Reporter<Activity = EpochCert>>(mut self, mut orchestrator: O) {
+        while let Some(message) = self.mailbox.next().await {
+            match message {
+                Message::Genesis { epoch, response } => {
+                    // Case: Genesis.
+                    if epoch == GENESIS_ROUND.epoch() {
+                        let _ = response.send(GENESIS_BLOCK.commitment());
+                        continue;
+                    }
+
+                    // Case: Non-genesis.
+                    let height = epoch::get_last_height(epoch - 1);
```

Comment:
I wonder if we want to give this last block a special name, "last height" can mean a lot of things.

Reply:
I suppose using some exotic name runs the risk of just being nebulous.

Reply:
Fixed in https://github.com/commonwarexyz/monorepo/pull/1460/commits/531503ac53ff61c9e972a313c1d8786b28588419